modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/OnflightMessageTracker.java

```diff
@@ -878,4 +906,70 @@ public void dumpMessageStatusToFile(File fileToWrite) throws AndesException {
         }
     }
 
+    /**
+     * Concurrent List like implementation.
+     * This implementation will create thread safe LinkedList.
+     * Two threads won't be able to execute same operation simultaneously
+     * with this LinkedList.
+     *
+     * @param <T> Type of list items
+     *
+     */
+    private static class ConcurrentTrackingList<T> {
```

Comment:
The difference between synchronizedList and this data structure is we have used Read-write locks instead of synchronized block. 

if you look at the usages of contains method it is used in org.wso2.andes.kernel.OnflightMessageTracker#addMessageToSendingTracker. This method will be called per each message send operation. In topics this send operations can happen parallely. Therefore if we had a normal synchonized block, only one message can pass at a time. But if we have read lock all contains can execute parallely when there is no writes. 

Therefore I think there is a benefit having this data structure instead of the synchronizedList.